lib/ansible/modules/identity/keycloak/keycloak_clienttemplate.py

```diff
@@ -0,0 +1,386 @@
+        required: false
+        choices: ['present', 'absent']
+        default: 'present'
+
+    id:
+        description:
+            - Id of client template to be worked on. This is usually a UUID.
+        required: false
+
+    name:
+        description:
+            - Name of the client template
+        required: false
+
+    description:
+        description:
+            - Description of the client template in Keycloak
+        required: false
+
+    bearer_only:
+        description:
+            - The access type of this client template is bearer-only.
+              This is 'bearerOnly' in the Keycloak REST API.
+        required: false
+
+    consent_required:
+        description:
+            - If enabled, users have to consent to client access.
+              This is 'consentRequired' in the Keycloak REST API.
+        required: false
+
+    standard_flow_enabled:
+        description:
+            - Enable standard flow for this client template or not (OpenID connect).
+              This is 'standardFlowEnabled' in the Keycloak REST API.
+        required: false
+
+    implicit_flow_enabled:
+        description:
+            - Enable implicit flow for this client template or not (OpenID connect).
+              This is 'implicitFlowEnabled' in the Keycloak REST API.
+        required: false
+
+    direct_access_grants_enabled:
+        description:
+            - Are direct access grants enabled for this client template or not (OpenID connect).
+              This is 'directAccessGrantsEnabled' in the Keycloak REST API.
+        required: false
+
+    service_accounts_enabled:
+        description:
+            - Are service accounts enabled for this client template or not (OpenID connect).
+              This is 'serviceAccountsEnabled' in the Keycloak REST API.
+        required: false
+
+    public_client:
+        description:
+            - Is the access type for this client template public or not.
+              This is 'publicClient' in the Keycloak REST API.
+        required: false
+
+    frontchannel_logout:
+        description:
+            - Is frontchannel logout enabled for this client template or not.
+              This is 'frontchannelLogout' in the Keycloak REST API.
+        required: false
+
+    protocol:
+        description:
+            - Type of client template (either C(openid-connect) or C(saml).
+        required: false
+        choices: ['openid-connect', 'saml']
+
+    full_scope_allowed:
+        description:
+            - Is the "Full Scope Allowed" feature set for this client template or not.
+              This is 'fullScopeAllowed' in the Keycloak REST API.
+        required: false
+
+    protocol_mappers:
+        description:
+            - a list of dicts defining protocol mappers for this client template.
+              An example of one is given in the examples section.
+              This is 'protocolMappers' in the Keycloak REST API.
+        required: false
```

Comment:
Can we make `protocol_mappers` more explicit?  Right now it is an arbitrary dict, without direct documentation, and without any specific validation.

This argument does not meet our new guidelines:  See number 5 in http://docs.ansible.com/ansible/devel/dev_guide/developing_modules.html#should-you-develop-a-module

Reply:
This will be difficult to do completely; while I can add some validation for the fields specified in the API documentation one level down, the Keycloak REST API documentation does not offer any documentation on valid fields for the "config" map therein. I can probably reverse engineer commonly used fields, but I would probably not be able to make it exhaustive, especially since some of it will depend on API version.

Is there leeway with doing such a best-effort approach? (i.e. documenting known entries, validating what is specified in the API docs, but also accepting key/value pairs that I may not have considered; I am not a developer on Keycloak / Red Hat SSO, merely a user/consumer of the API and do not have complete insight into the product and all its options.

Reply:
I have added the checks for one level down, and added documentation for those fields.

The config-map is mapper-specific, and mappers can be extended by the end-user; it would be counterproductive to disallow any fields.
It may be possible to validate at least some of the protocolMapper configs, however. The fields are not documented anywhere other than by appearing in the Keycloak sourcecode and through dumping configs from the API, however.